From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: redhoodsu <redhoodsu@tencent.com>
Date: Wed, 8 Dec 2021 20:24:11 +0800
Subject: add_chii_app.patch


diff --git a/BUILD.gn b/BUILD.gn
index 6462647da87dcc780717e5b52f70e06660f63b5e..d64ec56d9b9cdce680f4956ff4931c5889245807 100644
--- a/BUILD.gn
+++ b/BUILD.gn
@@ -64,6 +64,7 @@ application_templates = [ "front_end/entrypoint_template.html" ]
 
 devtools_applications = [
   "devtools_app",
+  "chii_app",
   "inspector",
   "js_app",
   "ndb_app",
diff --git a/config/gni/all_devtools_files.gni b/config/gni/all_devtools_files.gni
index 00cf41d5785622de8f3e94f1ded4086e3b6bfcdf..855599eb6bfddb691f354aae58b5f36caabc1fbe 100644
--- a/config/gni/all_devtools_files.gni
+++ b/config/gni/all_devtools_files.gni
@@ -7,6 +7,7 @@ import("../../scripts/build/ninja/vars.gni")
 all_devtools_files = [
   "front_end/entrypoints/shell/shell.js",
   "front_end/entrypoints/devtools_app/devtools_app.js",
+  "front_end/entrypoints/chii_app/chii_app.js",
   "front_end/entrypoints/inspector/inspector.js",
   "front_end/entrypoints/js_app/js_app.js",
   "front_end/entrypoints/ndb_app/ndb_app.js",
diff --git a/config/gni/devtools_grd_files.gni b/config/gni/devtools_grd_files.gni
index a0a80afb7c46836442dc5d7cb21b928aa9b66949..0904e4227c5f0786a7db225b438e2b9966843c03 100644
--- a/config/gni/devtools_grd_files.gni
+++ b/config/gni/devtools_grd_files.gni
@@ -162,6 +162,7 @@ grd_files_release_sources = [
   "front_end/core/sdk/sdk.js",
   "front_end/device_mode_emulation_frame.html",
   "front_end/devtools_app.html",
+  "front_end/chii_app.html",
   "front_end/devtools_compatibility.js",
   "front_end/emulated_devices/optimized/MotoG4-landscape.avif",
   "front_end/emulated_devices/optimized/MotoG4-portrait.avif",
@@ -206,6 +207,8 @@ grd_files_release_sources = [
   "front_end/entrypoints/device_mode_emulation_frame/device_mode_emulation_frame.js",
   "front_end/entrypoints/devtools_app/devtools_app-meta-files.js",
   "front_end/entrypoints/devtools_app/devtools_app.js",
+  "front_end/entrypoints/chii_app/chii_app-meta-files.js",
+  "front_end/entrypoints/chii_app/chii_app.js",
   "front_end/entrypoints/formatter_worker/FormatterActions.js",
   "front_end/entrypoints/formatter_worker/formatter_worker-entrypoint.js",
   "front_end/entrypoints/formatter_worker/formatter_worker.js",
diff --git a/front_end/BUILD.gn b/front_end/BUILD.gn
index 49cac86317ec6c86a9c88e149c7f15d1321fcfe5..0efc2604b9c0606cedf6a0644b8abca19e570a52 100644
--- a/front_end/BUILD.gn
+++ b/front_end/BUILD.gn
@@ -17,6 +17,7 @@ group("front_end") {
     ":legacy_entrypoints",
     "Images",
     "entrypoints/device_mode_emulation_frame",
+    "entrypoints/chii_app",
     "entrypoints/devtools_app",
     "entrypoints/formatter_worker:worker_entrypoint",
     "entrypoints/heap_snapshot_worker:worker_entrypoint",
@@ -118,6 +119,7 @@ node_action("html_entrypoints") {
 
   _devtools_templates = [
     "devtools_app",
+    "chii_app",
     "inspector",
     "js_app",
     "ndb_app",
diff --git a/front_end/entrypoints/chii_app/BUILD.gn b/front_end/entrypoints/chii_app/BUILD.gn
new file mode 100644
index 0000000000000000000000000000000000000000..afd6ad676bef43ee77886cb0ff5fac005d386412
--- /dev/null
+++ b/front_end/entrypoints/chii_app/BUILD.gn
@@ -0,0 +1,63 @@
+# Copyright 2021 The Chromium Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+import("../../../scripts/build/ninja/devtools_entrypoint.gni")
+import("../../../third_party/typescript/typescript.gni")
+import("../visibility.gni")
+
+devtools_entrypoint("chii_app-meta-files") {
+  entrypoint = "chii_app-meta-files.ts"
+
+  deps = [
+    "../../Images:optimize_images",
+    "../../core/dom_extension:bundle",
+    "../../core/sdk:meta",
+    "../../core/sdk:meta",
+    "../../entrypoints/main:meta",
+    "../../entrypoints/main:meta",
+    "../../entrypoints/startup:bundle",
+    "../../models/logs:meta",
+    "../../models/persistence:meta",
+    "../../panels/application:meta",
+    "../../panels/console:meta",
+    "../../panels/console_counters:meta",
+    "../../panels/input:meta",
+    "../../panels/protocol_monitor:meta",
+    "../../panels/settings:meta",
+    "../../panels/sources:meta",
+    "../../ui/legacy/components/object_ui:meta",
+    "../../ui/legacy/components/perf_ui:meta",
+    "../../ui/legacy/components/quick_open:meta",
+    "../../ui/legacy/components/source_frame:meta",
+    "../startup:bundle",
+    "../../entrypoints/inspector_main:meta",
+    "../../panels/accessibility:meta",
+    "../../panels/application:meta",
+    "../../panels/browser_debugger:meta",
+    "../../panels/developer_resources:meta",
+    "../../panels/elements:meta",
+    "../../panels/emulation:meta",
+    "../../panels/help:meta",
+    "../../panels/issues:meta",
+    "../../panels/layer_viewer:meta",
+    "../../panels/mobile_throttling:meta",
+    "../../panels/network:meta",
+    "../../panels/sensors:meta",
+  ]
+
+  visibility = [ ":*" ]
+
+  visibility += devtools_entrypoints_visibility
+}
+
+ts_library("chii_app") {
+  no_emit = true
+  sources = [ "chii_app.js" ]
+
+  deps = [
+    ":chii_app-meta-files",
+    "../../entrypoints/startup:bundle",
+    "../main:bundle",
+    "../shell",
+  ]
+}
diff --git a/front_end/entrypoints/chii_app/chii_app-meta-files.ts b/front_end/entrypoints/chii_app/chii_app-meta-files.ts
new file mode 100644
index 0000000000000000000000000000000000000000..053d2a435656b42b961017e5184dc4afc4fea55a
--- /dev/null
+++ b/front_end/entrypoints/chii_app/chii_app-meta-files.ts
@@ -0,0 +1,39 @@
+// Copyright 2021 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import '../../Images/Images.js';
+import '../startup/startup.js';
+import '../../core/root/root-legacy.js';
+import '../../core/platform/platform.js';
+import '../../core/dom_extension/dom_extension.js';
+
+import '../../panels/sources/sources-meta.js';
+import '../../panels/console/console-meta.js';
+import '../../panels/input/input-meta.js';
+import '../../panels/settings/settings-meta.js';
+import '../../panels/protocol_monitor/protocol_monitor-meta.js';
+import '../../models/persistence/persistence-meta.js';
+import '../../models/logs/logs-meta.js';
+import '../main/main-meta.js';
+import '../../ui/legacy/components/perf_ui/perf_ui-meta.js';
+import '../../ui/legacy/components/quick_open/quick_open-meta.js';
+import '../../core/sdk/sdk-meta.js';
+import '../../ui/legacy/components/source_frame/source_frame-meta.js';
+import '../../panels/console_counters/console_counters-meta.js';
+import '../../ui/legacy/components/object_ui/object_ui-meta.js';
+import '../main/main.js';
+
+import '../../panels/elements/elements-meta.js';
+import '../../panels/browser_debugger/browser_debugger-meta.js';
+import '../../panels/network/network-meta.js';
+import '../../panels/emulation/emulation-meta.js';
+import '../../panels/sensors/sensors-meta.js';
+import '../../panels/accessibility/accessibility-meta.js';
+import '../../panels/developer_resources/developer_resources-meta.js';
+import '../inspector_main/inspector_main-meta.js';
+import '../../panels/application/application-meta.js';
+import '../../panels/issues/issues-meta.js';
+import '../../panels/help/help-meta.js';
+import '../../panels/mobile_throttling/mobile_throttling-meta.js';
+import '../../panels/layer_viewer/layer_viewer-meta.js';
diff --git a/front_end/entrypoints/chii_app/chii_app.js b/front_end/entrypoints/chii_app/chii_app.js
new file mode 100644
index 0000000000000000000000000000000000000000..360df402c7fd52ceba0c01f93c6f2f06c06e7694
--- /dev/null
+++ b/front_end/entrypoints/chii_app/chii_app.js
@@ -0,0 +1,10 @@
+// Copyright 2018 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+import './chii_app-meta-files.js';
+
+import * as Main from '../main/main.js';
+import * as Startup from '../startup/startup.js';
+
+new Main.MainImpl.MainImpl();
+Startup.RuntimeInstantiator.startApplication('devtools_app');
diff --git a/front_end/entrypoints/chii_app/chii_app.json b/front_end/entrypoints/chii_app/chii_app.json
new file mode 100644
index 0000000000000000000000000000000000000000..58cfd4fb2212dfc8618d8a72ee3649d18176dfaf
--- /dev/null
+++ b/front_end/entrypoints/chii_app/chii_app.json
@@ -0,0 +1,4 @@
+{
+  "modules" : [
+  ]
+}
diff --git a/front_end/entrypoints/inspector_main/BUILD.gn b/front_end/entrypoints/inspector_main/BUILD.gn
index bcfd805139b43ead7f70fef9a002dbbaf8b046ab..b45549de3f1e13dd4ee60b431e740cf83071f38a 100644
--- a/front_end/entrypoints/inspector_main/BUILD.gn
+++ b/front_end/entrypoints/inspector_main/BUILD.gn
@@ -62,5 +62,5 @@ devtools_entrypoint("meta") {
 
   deps = [ ":bundle" ]
 
-  visibility = [ "../devtools_app/*" ]
+  visibility = [ "../devtools_app/*", "../chii_app/*" ]
 }
diff --git a/front_end/entrypoints/main/BUILD.gn b/front_end/entrypoints/main/BUILD.gn
index dbb6dfd7ee313f95cb47a8ed397444228a2a53d6..cded2443145d5248f16b5cd364e12b778e8de0ae 100644
--- a/front_end/entrypoints/main/BUILD.gn
+++ b/front_end/entrypoints/main/BUILD.gn
@@ -45,6 +45,7 @@ devtools_entrypoint("bundle") {
     ":*",
     "../../../test/unittests/front_end/entrypoints/missing_entrypoints/*",
     "../devtools_app:*",
+    "../chii_app:*",
     "../js_app:*",
     "../ndb_app:*",
     "../node_app:*",
@@ -70,5 +71,5 @@ devtools_entrypoint("meta") {
     "../../entrypoints/inspector_main:bundle",
   ]
 
-  visibility = [ "../shell/*" ]
+  visibility = [ "../shell/*", "../chii_app/*" ]
 }
diff --git a/front_end/entrypoints/shell/BUILD.gn b/front_end/entrypoints/shell/BUILD.gn
index 8013ba057685464da74930150e24367830fc27cd..706b61686434a7e34e1a3936f8e0636638e488a5 100644
--- a/front_end/entrypoints/shell/BUILD.gn
+++ b/front_end/entrypoints/shell/BUILD.gn
@@ -39,6 +39,7 @@ devtools_entrypoint("shell") {
   visibility = [
     "../..:*",
     "../devtools_app:*",
+    "../chii_app:*",
     "../js_app:*",
     "../ndb_app:*",
     "../node_app:*",
